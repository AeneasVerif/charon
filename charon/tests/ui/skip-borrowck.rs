//@ charon-args=--skip-borrowck
fn choose<'a, T>(b: bool, x: &'a mut T, y: &'a mut T) -> &'a mut T {
    if b {
        x
    } else {
        y
    }
}

pub fn choose_test() {
    let mut x = 0;
    let mut y = 0;
    let z = choose(true, &mut x, &mut y);
    *z += 1;
    assert!(*z == 1);
    // drop(z)
    assert!(x == 1);
    assert!(y == 0);
    assert!(*z == 1); // z is not valid anymore
}
