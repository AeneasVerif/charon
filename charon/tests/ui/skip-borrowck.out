# Final LLBC before serialization:

// Full name: core::marker::MetaSized
#[lang_item("meta_sized")]
pub trait MetaSized<Self>

// Full name: core::marker::Sized
#[lang_item("sized")]
pub trait Sized<Self>
{
    parent_clause0 : [@TraitClause0]: MetaSized<Self>
    non-dyn-compatible
}

// Full name: test_crate::choose
fn choose<'a, T>(@1: bool, @2: &'a mut (T), @3: &'a mut (T)) -> &'a mut (T)
where
    [@TraitClause0]: Sized<T>,
{
    let @0: &'_ mut (T); // return
    let b@1: bool; // arg #1
    let x@2: &'_ mut (T); // arg #2
    let y@3: &'_ mut (T); // arg #3
    let @4: &'_ mut (T); // anonymous local
    let @5: &'_ mut (T); // anonymous local
    let @6: bool; // anonymous local
    let @7: &'_ mut (T); // anonymous local

    storage_live(@7)
    storage_live(@4)
    storage_live(@5)
    storage_live(@6)
    @6 := copy (b@1)
    if move (@6) {
        storage_live(@7)
        @7 := &mut *(x@2)
        @5 := &mut *(@7)
        storage_dead(@7)
    }
    else {
        @5 := &mut *(y@3)
    }
    @4 := &mut *(@5)
    storage_dead(@6)
    @0 := &mut *(@4)
    storage_dead(@5)
    storage_dead(@4)
    return
}

// Full name: test_crate::choose_test
pub fn choose_test()
{
    let @0: (); // return
    let x@1: i32; // local
    let y@2: i32; // local
    let z@3: &'_ mut (i32); // local
    let @4: &'_ mut (i32); // anonymous local
    let @5: &'_ mut (i32); // anonymous local
    let @6: &'_ mut (i32); // anonymous local
    let @7: &'_ mut (i32); // anonymous local
    let @8: i32; // anonymous local
    let @9: (); // anonymous local
    let @10: bool; // anonymous local
    let @11: i32; // anonymous local
    let @12: (); // anonymous local
    let @13: bool; // anonymous local
    let @14: i32; // anonymous local
    let @15: (); // anonymous local
    let @16: bool; // anonymous local
    let @17: i32; // anonymous local
    let @18: (); // anonymous local
    let @19: bool; // anonymous local
    let @20: i32; // anonymous local

    storage_live(@8)
    storage_live(x@1)
    x@1 := const (0 : i32)
    storage_live(y@2)
    y@2 := const (0 : i32)
    storage_live(z@3)
    storage_live(@4)
    storage_live(@5)
    @5 := &mut x@1
    @4 := &two-phase-mut *(@5)
    storage_live(@6)
    storage_live(@7)
    @7 := &mut y@2
    @6 := &two-phase-mut *(@7)
    z@3 := choose<'_, i32>[Sized<i32>](const (true), move (@4), move (@6))
    storage_dead(@6)
    storage_dead(@4)
    storage_dead(@7)
    storage_dead(@5)
    @8 := copy (*(z@3)) panic.+ const (1 : i32)
    *(z@3) := move (@8)
    storage_live(@9)
    storage_live(@10)
    storage_live(@11)
    @11 := copy (*(z@3))
    @10 := move (@11) == const (1 : i32)
    if move (@10) {
    }
    else {
        storage_dead(@11)
        panic(core::panicking::panic)
    }
    storage_dead(@11)
    storage_dead(@10)
    storage_dead(@9)
    storage_live(@12)
    storage_live(@13)
    storage_live(@14)
    // drop(z)
    @14 := copy (x@1)
    @13 := move (@14) == const (1 : i32)
    if move (@13) {
    }
    else {
        storage_dead(@14)
        panic(core::panicking::panic)
    }
    storage_dead(@14)
    storage_dead(@13)
    storage_dead(@12)
    storage_live(@15)
    storage_live(@16)
    storage_live(@17)
    @17 := copy (y@2)
    @16 := move (@17) == const (0 : i32)
    if move (@16) {
    }
    else {
        storage_dead(@17)
        panic(core::panicking::panic)
    }
    storage_dead(@17)
    storage_dead(@16)
    storage_dead(@15)
    storage_live(@18)
    storage_live(@19)
    storage_live(@20)
    @20 := copy (*(z@3))
    @19 := move (@20) == const (1 : i32)
    if move (@19) {
    }
    else {
        storage_dead(@20)
        panic(core::panicking::panic)
    }
    storage_dead(@20)
    storage_dead(@19)
    storage_dead(@18)
    @0 := ()
    storage_dead(z@3)
    storage_dead(y@2)
    storage_dead(x@1)
    @0 := ()
    return
}



